README.md
# MSLoadingHUD

A simple swift for your iOS view controllers to easily show a loading HUD with or without a loading message.

Please note that MSLoadingHUD uses semantic versioning.

[![Version](https://img.shields.io/cocoapods/v/MSLoadingHUD.svg?style=flat)](https://cocoapods.org/pods/MSLoadingHUD)
[![License](https://img.shields.io/cocoapods/l/MSLoadingHUD.svg?style=flat)](https://cocoapods.org/pods/MSLoadingHUD)
[![Platform](https://img.shields.io/cocoapods/p/MSLoadingHUD.svg?style=flat)](https://cocoapods.org/pods/MSLoadingHUD)

## Example

To run the example project, clone the repo, and run `pod install` from the Example directory first.

## Requirements

## Installation


### Swift Package Manager

Add the following dependency to your Package.swift:

```swift
dependencies: [
    .package(url: "https://github.com/marcosiino/MSLoadingHUD.git", .upToNextMajor(from: "0.2.0"))
]
```

Or add the dependency to your app using Xcode: File => Swift Packages => Add Package Dependency... and type the git repo url: https://github.com/marcosiino/MSLoadingHUD.git

### CocoaPods

MSLoadingHUD is available through [CocoaPods](https://cocoapods.org). To install
it, simply add the following line to your Podfile:

```ruby
pod 'MSLoadingHUD'
```

## Usage

Add the Loadable trait to your view controller by declaring its conformance to the Loadable protocol:

```swift
class MyViewController: UIViewController, Loadable { ...
```

Then you can use the `showLoadingHUD` and `hideLoadingHUD` methods within your view controller's class to show and hide the loading HUD. The first method accepts an optional `loadingMessage` argument, which if it's specified, a message will be shown below the loading indicator.

Calling `showLoadingHUD` will cause any previous presented loading HUD to be hide before presenting the new one.

That's it!

## Author

Marco Siino, siino.marco [**] gmail.com

## License

MSLoadingHUD is available under the MIT license. See the LICENSE file for more info.


CHANGELOG.md
# Changelog

## 0.3.1 - 2020-12-06

### Fixes

- Fixed a problem which caused the loading hud not being hide in some cases


## Ver. 0.3.0 - 2020-12-06

### Changed

- showLoadingHUD method now automatically hides any previous shown loading huds

- An optional completion closure has been added to both showLoadingHUD and hideLoadingHUD methods, which informs when the operation is completed. 

- The loading box background color is now opaque 

## Ver. 0.2.0 - 2020-12-04

### Added
- Added Swift Package Manager compatibility

### Fixes
- Increased loading indicator container background opacity for better visibility


## Ver. 0.1.0 - 2020-12-04
Released the first version of this library with the basic features. 


Example/Tests/Tests.swift
import XCTest
import MSLoadingHUD

class Tests: XCTestCase {
    
    override func setUp() {
        super.setUp()
        // Put setup code here. This method is called before the invocation of each test method in the class.
    }
    
    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
        super.tearDown()
    }
    
    func testExample() {
        // This is an example of a functional test case.
        XCTAssert(true, "Pass")
    }
    
    func testPerformanceExample() {
        // This is an example of a performance test case.
        self.measure() {
            // Put the code you want to measure the time of here.
        }
    }
    
}


MSLoadingHUD/Classes/Loadable.swift
//
//  Loadable.swift
//  MSLoadingHUD
//
//  Created by Marco Siino on 04/12/2020.
//

import Foundation
import UIKit

/**
 Loadable is a trait you can add to any UIViewController to allow it to show and hide a simple loading HUD
 */
public protocol Loadable {
    func showLoadingHUD(loadingMessage: String)
    func hideLoadingHUD()
}

public extension Loadable where Self: UIViewController {
    
    /**
     Show the loading HUD
     - loadingMessage: The loading message. Default = "" (none)
     */
    func showLoadingHUD(loadingMessage: String = "") {
        //Shows the HUD
        func exec() {
            if let frontWindow = UIWindow.frontWindow {
                frontWindow.addSubview(LoadingView(frame: frontWindow.frame, loadingMessage: loadingMessage))
            }
        }
        
        //Hides any loading hud that may have been presented before first, then present the new one
        hideLoadingHUD()
        
        //Make sure the code is executed on the main thread
        if Thread.isMainThread {
            exec()
        }
        else {
            DispatchQueue.main.sync {
                exec()
            }
        }
    }
    
    /**
     Hides the presented loading HUD, if any
     */
    func hideLoadingHUD() {
        
        //Search the LoadingHUDViewController in any window hierarchy and remove it
        func exec() {
            for window in UIApplication.shared.windows {
                for v in window.subviews {
                    if v is LoadingView {
                        v.removeFromSuperview()
                    }
                }
            }
            
        }
        
        //Make sure to execute in main thread
        if Thread.isMainThread {
            exec()
        }
        else {
            DispatchQueue.main.sync {
                exec()
            }
        }
    }
}


Example/MSLoadingHUD/AppDelegate.swift
//
//  AppDelegate.swift
//  MSLoadingHUD
//
//  Created by Marco Siino on 12/04/2020.
//  Copyright (c) 2020 Marco Siino. All rights reserved.
//

import UIKit

@UIApplicationMain
class AppDelegate: UIResponder, UIApplicationDelegate {

    var window: UIWindow?


    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey: Any]?) -> Bool {
        // Override point for customization after application launch.
        return true
    }

    func applicationWillResignActive(_ application: UIApplication) {
        // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
        // Use this method to pause ongoing tasks, disable timers, and throttle down OpenGL ES frame rates. Games should use this method to pause the game.
    }

    func applicationDidEnterBackground(_ application: UIApplication) {
        // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later.
        // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
    }

    func applicationWillEnterForeground(_ application: UIApplication) {
        // Called as part of the transition from the background to the inactive state; here you can undo many of the changes made on entering the background.
    }

    func applicationDidBecomeActive(_ application: UIApplication) {
        // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
    }

    func applicationWillTerminate(_ application: UIApplication) {
        // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
    }


}



MSLoadingHUD/Classes/LoadingView.swift
//
//  LoadingView.swift
//  MSLoadingHUD
//
//  Created by Marco Siino on 04/12/2020.
//

import Foundation
import UIKit

class LoadingView: UIView {
    
    //Settingss
    private let loadingBoxMinWidth: CGFloat = 120.0
    private let loadingBoxMaxWidth: CGFloat = 220.0
    private let bgColor = UIColor.white
    private let foreColor = UIColor.black
    //*********
    
    private let loadingLabel: UILabel
    
    init(frame: CGRect, loadingMessage: String = "") {
        loadingLabel = UILabel(frame: CGRect.zero)
        super.init(frame: frame)
        
        loadingLabel.text = loadingMessage
        loadingLabel.textColor = foreColor.withAlphaComponent(0.7)
        loadingLabel.translatesAutoresizingMaskIntoConstraints = false
        loadingLabel.numberOfLines = 0
        loadingLabel.textAlignment = .center
        loadingLabel.font = UIFont.boldSystemFont(ofSize: 17.0)
        
        setup()
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    private func setup() {
        //Background color and effect
        backgroundColor = UIColor.black.withAlphaComponent(0.2)
        
        //Rounded corners container view
        let activityIndicatorContainer = UIView(frame: CGRect.zero)
        activityIndicatorContainer.translatesAutoresizingMaskIntoConstraints = false
        activityIndicatorContainer.backgroundColor = bgColor.withAlphaComponent(1.0)
        activityIndicatorContainer.layer.masksToBounds = true
        activityIndicatorContainer.layer.cornerRadius = 10.0
        addSubview(activityIndicatorContainer)
        
        //Activity indicator
        let activityIndicator = UIActivityIndicatorView()
        activityIndicator.translatesAutoresizingMaskIntoConstraints = false
        activityIndicator.color = foreColor
        activityIndicator.startAnimating()
        activityIndicatorContainer.addSubview(activityIndicator)
        
        //Adding text label to view hierarchy
        activityIndicatorContainer.addSubview(loadingLabel)
        
        //Constrain activity indicator container's constraints
        activityIndicatorContainer.centerXAnchor.constraint(equalTo: centerXAnchor).isActive = true
        activityIndicatorContainer.centerYAnchor.constraint(equalTo: centerYAnchor).isActive = true
        activityIndicatorContainer.widthAnchor.constraint(lessThanOrEqualToConstant: loadingBoxMaxWidth).isActive = true
        activityIndicatorContainer.widthAnchor.constraint(greaterThanOrEqualToConstant: loadingBoxMinWidth).isActive = true
        //Activity indicator's constraints
        activityIndicator.centerXAnchor.constraint(equalTo: activityIndicatorContainer.centerXAnchor).isActive = true
        activityIndicator.centerYAnchor.constraint(equalTo: activityIndicatorContainer.centerYAnchor).isActive = true
        //activityIndicator.topAnchor.constraint(equalTo: activityIndicatorContainer.topAnchor, constant: 45.0).isActive = true
        
        //Constraints for loading text label
        loadingLabel.leftAnchor.constraint(equalTo: activityIndicatorContainer.leftAnchor, constant: 16.0).isActive = true
        loadingLabel.rightAnchor.constraint(equalTo: activityIndicatorContainer.rightAnchor, constant: -16.0).isActive = true
        loadingLabel.topAnchor.constraint(equalTo: activityIndicator.bottomAnchor, constant: 16.0).isActive = true
        loadingLabel.bottomAnchor.constraint(equalTo: activityIndicatorContainer.bottomAnchor, constant: -16.0).isActive = true
        loadingLabel.heightAnchor.constraint(greaterThanOrEqualToConstant: 20.0).isActive = true
    }
    
    func setLoadingMessage(message: String) {
        loadingLabel.text = message
    }
}


Example/MSLoadingHUD/ViewController.swift
//
//  ViewController.swift
//  MSLoadingHUD
//
//  Created by Marco Siino on 12/04/2020.
//  Copyright (c) 2020 Marco Siino. All rights reserved.
//

import UIKit
import MSLoadingHUD

class ViewController: UIViewController, Loadable {

    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    @IBAction func simulateTask() {
        showLoadingHUD(loadingMessage: "Loading...")
        
        DispatchQueue.main.asyncAfter(deadline: .now() + 3.0, execute: {
            self.showLoadingHUD(loadingMessage: "Loading 2...")
            
            DispatchQueue.main.asyncAfter(deadline: .now() + 2.0, execute: {
                self.hideLoadingHUD()
            })
        })
        
        
    }

}



MSLoadingHUD/Classes/LoadingHUDViewController.swift
//
//  LoadingHUDViewController.swift
//  MSLoadingHUD
//
//  Created by Marco Siino on 04/12/2020.
//

import Foundation
import UIKit

class LoadingHUDViewController: UIViewController {
    
    //Settings
    private let loadingBoxMinWidth: CGFloat = 120.0
    private let loadingBoxMaxWidth: CGFloat = 220.0
    private let backgroundColor = UIColor.white
    private let foreColor = UIColor.black
    //*********
    
    private let loadingLabel: UILabel
    
    init(loadingMessage: String = "") {
        loadingLabel = UILabel(frame: CGRect.zero)
        super.init(nibName: nil, bundle: nil)
        
        modalPresentationStyle = .overFullScreen
        
        loadingLabel.text = loadingMessage
        loadingLabel.textColor = foreColor.withAlphaComponent(0.7)
        loadingLabel.translatesAutoresizingMaskIntoConstraints = false
        loadingLabel.numberOfLines = 0
        loadingLabel.textAlignment = .center
        loadingLabel.font = UIFont.boldSystemFont(ofSize: 17.0)
        
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        //Background color and effect
        view.backgroundColor = UIColor.black.withAlphaComponent(0.2)
        
        //Rounded corners container view
        let activityIndicatorContainer = UIView(frame: CGRect.zero)
        activityIndicatorContainer.translatesAutoresizingMaskIntoConstraints = false
        activityIndicatorContainer.backgroundColor = backgroundColor.withAlphaComponent(1.0)
        activityIndicatorContainer.layer.masksToBounds = true
        activityIndicatorContainer.layer.cornerRadius = 10.0
        view.addSubview(activityIndicatorContainer)
        
        //Activity indicator
        let activityIndicator = UIActivityIndicatorView()
        activityIndicator.translatesAutoresizingMaskIntoConstraints = false
        activityIndicator.color = foreColor
        activityIndicator.startAnimating()
        activityIndicatorContainer.addSubview(activityIndicator)
        
        //Adding text label to view hierarchy
        activityIndicatorContainer.addSubview(loadingLabel)
        
        //Constrain activity indicator container's constraints
        activityIndicatorContainer.centerXAnchor.constraint(equalTo: view.centerXAnchor).isActive = true
        activityIndicatorContainer.centerYAnchor.constraint(equalTo: view.centerYAnchor).isActive = true
        activityIndicatorContainer.widthAnchor.constraint(lessThanOrEqualToConstant: loadingBoxMaxWidth).isActive = true
        activityIndicatorContainer.widthAnchor.constraint(greaterThanOrEqualToConstant: loadingBoxMinWidth).isActive = true
        //Activity indicator's constraints
        activityIndicator.centerXAnchor.constraint(equalTo: activityIndicatorContainer.centerXAnchor).isActive = true
        activityIndicator.centerYAnchor.constraint(equalTo: activityIndicatorContainer.centerYAnchor).isActive = true
        //activityIndicator.topAnchor.constraint(equalTo: activityIndicatorContainer.topAnchor, constant: 45.0).isActive = true
        
        //Constraints for loading text label
        loadingLabel.leftAnchor.constraint(equalTo: activityIndicatorContainer.leftAnchor, constant: 16.0).isActive = true
        loadingLabel.rightAnchor.constraint(equalTo: activityIndicatorContainer.rightAnchor, constant: -16.0).isActive = true
        loadingLabel.topAnchor.constraint(equalTo: activityIndicator.bottomAnchor, constant: 16.0).isActive = true
        loadingLabel.bottomAnchor.constraint(equalTo: activityIndicatorContainer.bottomAnchor, constant: -16.0).isActive = true
        loadingLabel.heightAnchor.constraint(greaterThanOrEqualToConstant: 20.0).isActive = true
    
    }
    
    func setLoadingMessage(message: String) {
        loadingLabel.text = message
    }
    
    override func loadView() {
        view = UIView()
    }
}
